lib/IDE/CodeCompletion.cpp

```diff
@@ -1565,6 +1566,34 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
     return false;
   }
 
+  /// Returns \c true if \p TAD is usable as a first type of a requirement in
+  /// \c where clause for a context.
+  /// \p selfTy must be a \c Self type of the context.
+  static bool canBeUsedAsRequirementFirstType(Type selfTy, TypeAliasDecl *TAD) {
+    auto T = TAD->getDeclaredInterfaceType();
+    auto subMap = selfTy->getMemberSubstitutionMap(TAD->getParentModule(), TAD);
+    T = T.subst(subMap)->getCanonicalType();
+
+    ArchetypeType *archeTy = T->getAs<ArchetypeType>();
+    if (!archeTy)
+      return false;
+    archeTy = archeTy->getRoot();
+
+    // For protocol, the 'archeTy' should match with the 'baseTy' which is the
+    // dynamic 'Self' type of the protocol.
+    if (archeTy->isEqual(selfTy))
+      return true;
+
+    // For nominal decls, 'archTy' should be one of the generic parameters.
+    if (selfTy->castTo<BoundGenericType>() &&
```

Comment:
`castTo` is not allowed to return nullptr; did you mean to use `getAs`?

Reply:
You're right. I meant `getAs`. But I reconsider it, and updated to find `archeTy` from `selfTy`.